BugFix: Added support for net6.0 for J2N.Randomizer and fixed several members that behaved differently than the JDK

NightOwl888 opened the pull request:
This packages `J2N.Randomizer` into its own NuGet package to re-add support for `net6.0` and higher when J2N is targeting `netstandard2.1`. This is a problem for testing using `net6.0`, which we recently changed on some of our consuming projects from `net5.0`. System.Random added support for `NextInt64()` and `NextSingle()` in `net6.0`, but since users of the J2N `netstandard2.1` target don't have access to those APIs when they are consuming it with `net6.0`, it caused them to call the BCL implementation instead of ours, which doesn't suit our purposes in many ways:

1. Random values are not the same across target frameworks
2. Random values are not the same across operating systems and hardware
3. The same random seed doesn't produce the same results on multiple instances (due to the fact that we don't seed the base class from `J2N.Randomizer`, as we are supposed to be overriding it)
4. The random values are quite different from those returned from the Java implementation of Random

This behavior is unacceptable for our purposes and the best option for fixing it is to continue supporting `net6.0`. But doing that across all of J2N would interfere with our ability to test `netstandard2.1`, so we are only adding support for `J2N.Randomizer` by moving it to its own package. This package is a dependency of J2N, but only for the `netstandard2.1` target. All other target frameworks will still have a single assembly and single NuGet package.

This bug was introduced in #122 when we dropped support for `net6.0`.

-----------------------

This PR also makes breaking behavioral changes to the way `J2N.Randomizer` functions. Specifically, it changes the values that are returned to exactly match the behavior of the JDK bit by bit (with the exception of the `NextGaussian()` method, which returns a more accurate value than in Java that is not rounded exactly the same way).

- The `NextInt64(long)` and `NextInt64(long, long)` overloads were implemented to match the JDK 22. Older JDKs did not provide implementations of these, including Apache Harmony - the JDK that we ported. Implementing these covers a gap in the implementation so we always override the BCL behavior.
- The `NextBytes(Span<byte>)` overload was also implemented to override the default BCL behavior and to align with our efforts to support System.Memory types.
- The `Next(int)` and `Next(int, int)` implementations were changed to match the results from the JDK based on the same seed.
- Tests were added to ensure we have matching binary behavior for the same seed on all public APIs that return random results.

For Lucene.Net.TestFramework users, any random seeds that have been reported to be able to reproduce a test failure may no longer do so after upgrading to this version of `Randomizer`. Unfortunately, although the results were repeatable, the implementation we ported from Apache Harmony doesn't quite match the behavior of the JDK. This constitutes a breaking behavioral change, but there are no breaking binary changes to the API.

Note that the [`Random.Shuffle()` overloads](https://learn.microsoft.com/en-us/dotnet/api/system.random.shuffle?view=net-9.0) and [`Random.GetItems()` overloads](https://learn.microsoft.com/en-us/dotnet/api/system.random.getitems?view=net-9.0) cannot be overridden and rely on the behavior of the value returned by `Random.Next()` (which is overridden). So, while not technically broken, these methods do not behave like the Java implementation. Alternatives to use for Java compatibility are the `J2N.Collections.Generic.ListExtensions.Shuffle()` extension method and the `RandomPicks` class in `RandomizedTesting.Generators`, although these alternatives don't support System.Memory types at the time of this writing.

paulirwin commented on src/J2N.Randomizer/J2N.Randomizer.csproj:
```diff
@@ -0,0 +1,56 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>net9.0;net8.0;net6.0;netstandard2.1</TargetFrameworks>
```
If this project is only included for netstandard2.1, are the other target frameworks needed here? I don't see a reason to reject this PR over this, but just curious.

NightOwl888 replied:
Well, at minimum we will need `net6.0` to ensure the Random overloads exist for consumers in .NET Core. Without that, the fix doesn't work. It is also important to override any additional virtual methods that Microsoft adds in the future.

Granted, `net8.0` and `net9.0` will have a type conflict if both packages are installed, but it is not very likely to happen and we explicitly state this package isn't intended for direct consumption.

paulirwin reviewed (approved):
Comment left with a question, but otherwise looks good. I am also not a random algorithm expert, but I don't see anything outright wrong here.

NightOwl888 commented:
@paulirwin

I added some tests from JDK 22 that target each branch in the `Next(int)`, `Next(int, int)`, `NextInt64(long)` and `NextInt64(long, long)` overloads.

It turns out that `nextInt(int)` wasn't added until Java 9 . `nextInt(int, int)`, `nextLong(long)`, and `nextLong(long, long)` were added in Java 17, but they were put onto an interface rather than the main API. There was a breaking change in Java 21 where `nextInt(int, int)` and `nextLong(long)` were changed to return a different value for a given set of inputs. The other overloads didn't break binary compatibility going back to at least Java 8.

With that in mind, I went ahead and added an implementation that is equivalent to the Java 21 behavior. This is not a direct port, but a cleanroom implementation with different variable names and control flow logic than the JDK to avoid any licensing concerns. Unfortunately, Harmony was discontinued long before these overloads were a thing, so behaving like Java 17 or like Java 21 to stay binary compatible are the only options.

I would have rather not implemented these since we don't use them, but since `System.Random` added these overloads in .NET 6, we either need to:

- Provide a JDK compatible implementation
- Throw a `NotSupportedException`
- Leave the default BCL behavior

Leaving the default BCL behavior isn't a good option because it doesn't use our seed, so making it both repeatable and settable is not possible, but it is what happens now in .NET 6+ because these overloads were added to `System.Random` after `Randomizer` existed. Throwing an exception means that any callers currently using it will have a breaking change to their code with no workaround.

paulirwin commented:
Good catch, and I agree with your approach.
